pkg/storage/image.go

```diff
@@ -176,145 +178,106 @@ func (svc *imageService) makeRepoDigests(knownRepoDigests, tags []string, imageI
+		return imageCacheItem{}, err
+	}
+	imageFull, err := ref.NewImage(systemContext)
+	if err != nil {
+		return imageCacheItem{}, err
+	}
+	defer imageFull.Close()
+	imageConfig, err := imageFull.OCIConfig()
+	if err != nil {
+		return imageCacheItem{}, err
+	}
+	return imageCacheItem{
+		user:         imageConfig.Config.User,
+		size:         size,
+		configDigest: configDigest,
+	}, nil
+}
+
+func (svc *imageService) appendCachedResult(systemContext *types.SystemContext, ref types.ImageReference, image *storage.Image, results []ImageResult, newImageCache imageCache) ([]ImageResult, error) {
+	var err error
+	svc.imageCacheLock.Lock()
+	cacheItem, ok := svc.imageCache[image.ID]
+	svc.imageCacheLock.Unlock()
+	if !ok {
+		cacheItem, err = svc.buildImageCacheItem(systemContext, ref, image)
+		if err != nil {
+			return results, err
+		}
+		if newImageCache == nil {
+			svc.imageCacheLock.Lock()
+			svc.imageCache[image.ID] = cacheItem
+			svc.imageCacheLock.Unlock()
+		} else {
+			newImageCache[image.ID] = cacheItem
+		}
+	} else if newImageCache != nil {
+		newImageCache[image.ID] = cacheItem
+	}
+
+	name, tags, digests := sortNamesByType(image.Names)
+	imageDigest, repoDigests := svc.makeRepoDigests(digests, tags, image.ID)
+	return append(results, ImageResult{
+		ID:           image.ID,
+		Name:         name,
+		RepoTags:     tags,
+		RepoDigests:  repoDigests,
+		Size:         cacheItem.size,
+		Digest:       imageDigest,
+		ConfigDigest: cacheItem.configDigest,
+		User:         cacheItem.user,
+	}), nil
+}
+
 func (svc *imageService) ListImages(systemContext *types.SystemContext, filter string) ([]ImageResult, error) {
-	results := []ImageResult{}
+	var results []ImageResult
 	if filter != "" {
+		// we never remove entries from cache unless unfiltered ListImages call is made. Is it safe?
 		ref, err := svc.getRef(filter)
 		if err != nil {
 			return nil, err
 		}
 		if image, err := istorage.Transport.GetStoreImage(svc.store, ref); err == nil {
-			var user string
-			var size *uint64
-			var configDigest digest.Digest
-			if cacheItem, ok := svc.imageCache[image.ID]; ok {
-				user, size, configDigest = cacheItem.user, cacheItem.size, cacheItem.configDigest
-			} else {
-				img, err := ref.NewImageSource(systemContext)
-				if err != nil {
-					return nil, err
-				}
-				size = imageSize(img)
-				configDigest, err = imageConfigDigest(img, nil)
-				img.Close()
-				if err != nil {
-					return nil, err
-				}
-				imageFull, err := ref.NewImage(systemContext)
-				if err != nil {
-					return nil, err
-				}
-				defer imageFull.Close()
-				imageConfig, err := imageFull.OCIConfig()
-				if err != nil {
-					return nil, err
-				}
-				user = imageConfig.Config.User
-				cacheItem := imageCacheItem{
-					user:         user,
-					size:         size,
-					configDigest: configDigest,
-				}
-				svc.imageCache[image.ID] = cacheItem
+			results, err = svc.appendCachedResult(systemContext, ref, image, []ImageResult{}, nil)
```

Comment:
This can pass `nil` instead of `[]ImageResult{}`.

Reply:
> This can pass `nil` instead of `[]ImageResult{}`.

@nalind, do you want to file a follow-up PR with this, or should we just leave it?

Reply:
Eh, personal preference, but not a blocker either way.